flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/source/metrics/SourceEnumeratorMetrics.java

```diff
@@ -0,0 +1,251 @@
+
+    /**
+     * Collection class for managing metrics of a table.
+     *
+     * <p>Metrics of table level are registered in its corresponding subgroup under the {@link
+     * SplitEnumeratorMetricGroup}.
+     */
+    public static class TableMetrics {
+        private AtomicInteger numSnapshotSplitsProcessed = new AtomicInteger(0);
+        private AtomicInteger numSnapshotSplitsRemaining = new AtomicInteger(0);
+        private AtomicInteger numSnapshotSplitsFinished = new AtomicInteger(0);
+        private volatile long snapshotStartTime = UNDEFINED;
+        private volatile long snapshotEndTime = UNDEFINED;
+
+        private Set<String> remainingSplitIds =
+                Collections.newSetFromMap(new ConcurrentHashMap<>());
+        private Set<String> processedSplitIds =
+                Collections.newSetFromMap(new ConcurrentHashMap<>());
+        private Set<String> finishedSplitIds = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+        public TableMetrics(
+                String databaseName, String schemaName, String tableName, MetricGroup parentGroup) {
+            databaseName = processNull(databaseName);
+            schemaName = processNull(schemaName);
+            tableName = processNull(tableName);
+            MetricGroup metricGroup =
+                    parentGroup
+                            .addGroup(DATABASE_GROUP_KEY, databaseName)
+                            .addGroup(SCHEMA_GROUP_KEY, schemaName)
+                            .addGroup(TABLE_GROUP_KEY, tableName);
+            metricGroup.gauge(
+                    NUM_SNAPSHOT_SPLITS_PROCESSED, () -> numSnapshotSplitsProcessed.intValue());
+            metricGroup.gauge(
+                    NUM_SNAPSHOT_SPLITS_REMAINING, () -> numSnapshotSplitsRemaining.intValue());
+            metricGroup.gauge(
+                    NUM_SNAPSHOT_SPLITS_FINISHED, () -> numSnapshotSplitsFinished.intValue());
+            metricGroup.gauge(SNAPSHOT_START_TIME, () -> snapshotStartTime);
+            metricGroup.gauge(SNAPSHOT_END_TIME, () -> snapshotEndTime);
+            snapshotStartTime = System.currentTimeMillis();
+        }
+
+        private String processNull(String name) {
+            if (StringUtils.isBlank(name)) {
+                // If null, convert to an empty string
+                return "";
+            }
+            return name;
+        }
+
+        public void addNewSplit(String newSplitId) {
+            if (!remainingSplitIds.contains(newSplitId)) {
+                remainingSplitIds.add(newSplitId);
+                numSnapshotSplitsRemaining.getAndAdd(1);
+                LOGGER.info("add remaining split: {}", newSplitId);
+            }
+        }
+
+        public void addNewSplits(List<String> newSplitIds) {
+            if (newSplitIds != null) {
+                for (String newSplitId : newSplitIds) {
+                    addNewSplit(newSplitId);
```

Comment:
Please do not add these splits one by one. We only need to add them with addAll and size.